Tutorial for Egress (#1259)
kyma-project/istio: docs/user/tutorials/01-50-send-requests-using-egress.md

# Send requests using Istio Egress

## Prerequisites

* You have the Istio module added.
* To use CLI instruction, you must install [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl)

>   and [curl](https://curl.se/). Alternatively, you can use Kyma dashboard.

Reviewer: I think it does not make sense here, as there are no UI steps
```suggestion
  and [curl](https://curl.se/).
```

### Configuration

1. Export the following values as environment variables:

    ```bash
    export NAMESPACE={service-namespace}
    ```

2. Create a new namespace for the sample application.
    ```bash
    kubectl create ns $NAMESPACE
    kubectl label namespace $NAMESPACE istio-injection=enabled --overwrite
    ```

> 3. Enable additional sidecar logs to see egresGateway being used in requests:

Reviewer: ```suggestion
3. Enable additional sidecar logs to see egressGateway being used in requests:
```

        matchLabels:
          app: curl
      template:
        metadata:
          labels:
            app: curl
        spec:
          terminationGracePeriodSeconds: 0
          serviceAccountName: curl
          containers:
          - name: curl
            image: curlimages/curl
            command: ["/bin/sleep", "infinity"]
            imagePullPolicy: IfNotPresent
            volumeMounts:
            - mountPath: /etc/curl/tls
              name: secret-volume
          volumes:
          - name: secret-volume
            secret:
              secretName: curl-secret
              optional: true
    EOF
    ```

   Get the `curl` pod:
    ```bash
   export SOURCE_POD=$(kubectl get pod -n "$NAMESPACE" -l app=curl -o jsonpath={.items..metadata.name})
    ```

5. Define a `ServiceEntry`:
   
   ```bash
   kubectl apply -f - <<EOF
   apiVersion: networking.istio.io/v1
   kind: ServiceEntry
   metadata:
     name: kyma-project
     namespace: $NAMESPACE
   spec:
     hosts:
     - kyma-project.io
     ports:
     - number: 443
       name: tls
       protocol: TLS
     resolution: DNS
   EOF
   ```
   
   Verify that the `ServiceEntry` was created successfully:
   ```bash

>    kubectl exec -n "$NAMESPACE" "$SOURCE_POD" -c curl -- curl -sSL -o /dev/null -D - https://kyma-project.io

Reviewer: please add command output for with and without service entry (clarify what changes).
I see no difference with or without it.

   ```

6. Create an egress `Gateway`, `DestinationRule` and `VirtualService` to direct traffic:
   
   ```bash
   kubectl apply -f - <<EOF
   apiVersion: networking.istio.io/v1
   kind: Gateway
   metadata:
     name: istio-egressgateway
     namespace: ${NAMESPACE}
   spec:
     selector:
       istio: egressgateway
     servers:
     - port:
         number: 443
         name: tls
         protocol: TLS
       hosts:
       - kyma-project.io
       tls:
         mode: PASSTHROUGH
   ---
   apiVersion: networking.istio.io/v1
   kind: DestinationRule
   metadata:
     name: egressgateway-for-kyma-project
     namespace: ${NAMESPACE}
   spec:
     host: istio-egressgateway.istio-system.svc.cluster.local
     subsets:
     - name: kyma-project
   ---
   apiVersion: networking.istio.io/v1
   kind: VirtualService
   metadata:
     name: direct-kyma-project-through-egress-gateway
     namespace: ${NAMESPACE}

>    spec:

Reviewer: Might make sense to add something that actually shows that the VirtualService does something, e.g. some delay injection, or header modification etc.

Author: I don't think that's the purpose of this tutorial, we just want to show routing requests through egress

  hosts:
  - kyma-project.io
  gateways:
  - mesh
  - istio-egressgateway
  tls:
  - match:
    - gateways:
      - mesh
      port: 443
      sniHosts:
      - kyma-project.io
    route:
    - destination:
        host: istio-egressgateway.istio-system.svc.cluster.local
        subset: kyma-project
        port:
          number: 443
  - match:
    - gateways:
      - istio-egressgateway
      port: 443
      sniHosts:
      - kyma-project.io
    route:
    - destination:
        host: kyma-project.io
        port:
          number: 443
      weight: 100
EOF
```

Once again, send an HTTPS reqest:
```bash
kubectl exec -n "$NAMESPACE" "$SOURCE_POD" -c curl -- curl -sSL -o /dev/null -D - https://kyma-project.io

> ```

Reviewer: Please add expected command output here as well

Check egress gateway log:
```bash

> kubectl logs -l istio=egressgateway -n istio-system

Reviewer: Please add expected log, and what to look for

Reviewer: It probably makes sense to compare the logs without the routing through EG applied

# Send requests using Istio Egress

## Prerequisites

* Istio module with egress enabled.
* To use CLI instruction, you must install [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl)
  and [curl](https://curl.se/).

### Configuration

1. Export the following values as environment variables:

    ```bash
    export NAMESPACE={service-namespace}
    ```

2. Create a new namespace for the sample application.
    ```bash
    kubectl create ns $NAMESPACE
    kubectl label namespace $NAMESPACE istio-injection=enabled --overwrite
    ```

3. Make sure there is an Istio CR with egress enabled:
   ```bash
   kubectl apply -f - <<EOF
   apiVersion: operator.kyma-project.io/v1alpha2
   kind: Istio
   metadata:
     name: default
     namespace: kyma-system
     labels:
       app.kubernetes.io/name: default
   spec:
     components:
       egressGateway:
         enabled: true
   EOF
   ```

> 4. Enable additional sidecar logs to see egressGateway being used in requests:

Reviewer 2: ```suggestion
4. Enable additional sidecar logs to see the egress Gateway being used in requests:
```
tbh, I don't know if it should be egressGateway / egress gateway / or egress Gateway. I used the form 'egress Gateway' throughout the tutorial so it's written in the same way

    apiVersion: v1
    kind: Service
    metadata:
      name: curl
      namespace: ${NAMESPACE}
      labels:
        app: curl
        service: curl
    spec:
      ports:
      - port: 80
        name: http
      selector:
        app: curl
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: curl
      namespace: ${NAMESPACE}
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: curl
      template:
        metadata:
          labels:
            app: curl
        spec:
          terminationGracePeriodSeconds: 0
          serviceAccountName: curl
          containers:
          - name: curl
            image: curlimages/curl
            command: ["/bin/sleep", "infinity"]
            imagePullPolicy: IfNotPresent
            volumeMounts:
            - mountPath: /etc/curl/tls
              name: secret-volume
          volumes:
          - name: secret-volume
            secret:
              secretName: curl-secret
              optional: true
    EOF
    ```

   Export the name of the `curl` Pod:
    ```bash
   export SOURCE_POD=$(kubectl get pod -n "$NAMESPACE" -l app=curl -o jsonpath={.items..metadata.name})
    ```

> 6. Define a ServiceEntry to allow outbound traffic to the `kyma-project` domain and perform DNS resolution:

Reviewer: > to allow outbound traffic to the `kyma-project` domain and perform DNS resolution
this will generally be not true. This is true if the Istio outside access mode of the sidecar / mesh is set to mode [REGISTRY_ONLY](https://istio.io/latest/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services). In Kyma scenario we install Istio with ALLOW_ANY, so here the ServiceEntry has actually a different purpose: to be recognised as a service inside the mesh.
See: https://istiobyexample.dev/external-services/
